File: Quasics2019Code/CanDriveTrain/src/main/cpp/commands/TankDrive.cpp
/*----------------------------------------------------------------------------*/
/* Copyright (c) 2017-2018 FIRST. All Rights Reserved.                        */
/* Open Source Software - may be modified and shared by FRC teams. The code   */
/* must be accompanied by the FIRST BSD license file in the root directory of */
/* the project.                                                               */
/*----------------------------------------------------------------------------*/

#include "commands/TankDrive.h"
#include "subsystems/DriveBase.h"
#include "Robot.h"
#include "OI.h"

const float kDeadBandSize = 0.015;

TankDrive::TankDrive() {
  // Use Requires() here to declare subsystem dependencies
  // eg. Requires(Robot::chassis.get());
  Requires(Robot::driveBase.get());
}

// Called just before this Command runs the first time
void TankDrive::Initialize() {}

// Called repeatedly when this Command is scheduled to run
void TankDrive::Execute() {
  double leftStickReading = Robot::m_oi.GetLeftJoystick();
  double rightStickReading = Robot::m_oi.GetRightJoystick();

  // Software dead-band handling.
  if (std::abs(leftStickReading) < kDeadBandSize) {
    leftStickReading = 0.0;
  }
  if (std::abs(rightStickReading) < kDeadBandSize) {
    rightStickReading = 0.0;
  }

  Robot::driveBase->SetMotorSpeed(
      leftStickReading * mult,
      rightStickReading * mult);
}

// Make this return true when this Command no longer needs to run execute()
bool TankDrive::IsFinished() { return false; }

// Called once after isFinished returns true
void TankDrive::End() {
  Robot::driveBase->Stop();
}

// Called when another command which requires one or more of the same
// subsystems is scheduled to run
void TankDrive::Interrupted() {
  End();
}

File: Quasics2026Code/MaeTester2026-Java/src/main/java/frc/robot/commands/TankDrive.java
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

package frc.robot.commands;

import edu.wpi.first.wpilibj2.command.Command;
import frc.robot.subsystems.Drivebase;
import java.util.function.Supplier;

/**
 * Command providing support for driving the robot in "tank drive" mode.
 */
public class TankDrive extends Command {
  private final Drivebase m_drivebase;
  private final Supplier<Double> m_leftPowerFunction;
  private final Supplier<Double> m_rightPowerFunction;

  /** Creates a new TankDrive. */
  public TankDrive(Drivebase drivebase, Supplier<Double> leftPowerFunction,
      Supplier<Double> rightPowerFunction) {
    m_drivebase = drivebase;
    m_leftPowerFunction = leftPowerFunction;
    m_rightPowerFunction = rightPowerFunction;
    addRequirements(drivebase);
  }

  /** Updates the current drive settings. */
  private void updateSpeeds() {
    m_drivebase.setMotorSpeed(
        m_leftPowerFunction.get(), m_rightPowerFunction.get());
  }

  // Called when the command is initially scheduled.
  @Override
  public void initialize() {
    updateSpeeds();
  }

  // Called every time the scheduler runs while the command is scheduled.
  @Override
  public void execute() {
    updateSpeeds();
  }

  // Called once the command ends or is interrupted.
  @Override
  public void end(boolean interrupted) {
    m_drivebase.stop();
  }
}
